code/controllers/subsystem/zmimic.dm

```diff
@@ -0,0 +1,734 @@
+/datum/controller/subsystem/zmimic/proc/RebuildZState()
+	// suspend()
+	UNTIL(state == SS_IDLE)
+
+	calculate_zstack_limits()
+
+	for (var/zlev in 1 to world.maxz)
+		for (var/turf/T in block(locate(1, 1, zlev), locate(world.maxx, world.maxy, zlev)))
+			if (T.mz_flags & MZ_MIMIC_BELOW)
+				flush_z_state(T)
+				T.below = GetAbove(T)
+				T.above = GetBelow(T)
+				T.update_mimic()
+			CHECK_TICK
+	// wake()
+
+
+/datum/controller/subsystem/zmimic/fire(resumed = FALSE, no_mc_tick = FALSE)
+	if (!resumed)
+		qt_idex = 1
+		qo_idex = 1
+
+	MC_SPLIT_TICK_INIT(2)
+	if (!no_mc_tick)
+		MC_SPLIT_TICK
+
+	var/list/curr_turfs = queued_turfs
+	var/list/curr_ov = queued_overlays
+
+	while (qt_idex <= curr_turfs.len)
+		var/turf/T = curr_turfs[qt_idex]
+		curr_turfs[qt_idex] = null
+		qt_idex += 1
+
+		if (!isturf(T) || !(T.mz_flags & MZ_MIMIC_BELOW) || !T.z_queued)
+			if (no_mc_tick)
+				CHECK_TICK
+			else if (MC_TICK_CHECK)
+				break
+			continue
+
+		// If we're not at our most recent queue position, don't bother -- we're updating again later anyways.
+		if (T.z_queued > 1)
+			T.z_queued -= 1
+			multiqueue_skips_turf += 1
+			if (no_mc_tick)
+				CHECK_TICK
+			else if (MC_TICK_CHECK)
+				break
+			continue
+
+		// Z-Turf on the bottom-most level, just fake-copy space (or baseturf).
+		// It's impossible for anything to be on the synthetic turf, so ignore the rest of the ZM machinery.
+		if (!T.below)
+			flush_z_state(T)
+			if (T.mz_flags & MZ_MIMIC_BASETURF)
+				simple_appearance_copy(T, T.baseturfs, OPENTURF_MAX_PLANE)
+			// else
+				// TODO SSparallax skybox support.
+				// simple_appearance_copy(T, SSskybox.dust_cache["[((T.x + T.y) ^ ~(T.x * T.y) + T.z) % 25]"])
+
+			if (no_mc_tick)
+				CHECK_TICK
+			else if (MC_TICK_CHECK)
+				break
+			continue
+
+		if (!T.shadower)	// If we don't have a shadower yet, something has gone horribly wrong.
+			WARNING("Turf [T] at [T.x],[T.y],[T.z] was queued, but had no shadower.")
+			continue
+
+		T.z_generation += 1
+
+		// Get the bottom-most turf, the one we want to mimic.
+		var/turf/Td = T
+		while (Td.below)
+			Td = Td.below
+
+		// Depth must be the depth of the *visible* turf, not self.
+		var/turf_depth
+		turf_depth = T.z_depth = zlev_maximums[Td.z] - Td.z
+
+		var/t_target = OPENTURF_MAX_PLANE - turf_depth	// This is where the turf (but not the copied atoms) gets put.
+
+		// Turf is set to mimic baseturf, handle that and bail.
+		if (T.mz_flags & MZ_MIMIC_BASETURF)
+			flush_z_state(T)
+			simple_appearance_copy(T, T.baseturfs, t_target)
+
+			if (no_mc_tick)
+				CHECK_TICK
+			else if (MC_TICK_CHECK)
+				break
+			continue
+
+		// If we previously were MZ_MIMIC_BASETURF, there might be an orphaned proxy.
+		else if (T.mimic_underlay)
+			QDEL_NULL(T.mimic_underlay)
+
+
+		// Handle space parallax & starlight.
+		if (T.below.z_eventually_space)
+			T.z_eventually_space = TRUE
+			t_target = SPACE_PLANE
```

Comment:
I'm not certain that this is applicable to this ZM revision, but this probably should be:
```suggestion
			if ((T.below.mz_flags & MZ_MIMIC_OVERWRITE) || T.below.type == /turf/space)
						t_target = SPACE_PLANE
```

iirc this fixes an issue with glass on top of parallax/space.